src/View/SSViewer.php

```diff
@@ -57,210 +36,98 @@ class SSViewer implements Flushable
      */
-    protected static $topLevel = [];
+    protected bool $includeRequirements = true;
 
-    /**
-     * List of templates to select from
-     *
-     * @var array
-     */
-    protected $templates = null;
+    private TemplateEngine $templateEngine;
 
     /**
-     * Absolute path to chosen template file
-     *
-     * @var string
-     */
-    protected $chosen = null;
-
-    /**
-     * Templates to use when looking up 'Layout' or 'Content'
-     *
-     * @var array
-     */
-    protected $subTemplates = [];
-
-    /**
-     * @var bool
-     */
-    protected $includeRequirements = true;
-
-    /**
-     * @var TemplateParser
-     */
-    protected $parser;
-
-    /**
-     * @var CacheInterface
-     */
-    protected $partialCacheStore = null;
-
-    /**
-     * @param string|array $templates If passed as a string with .ss extension, used as the "main" template.
+     * @param string|array $templates If passed as a string, used as the "main" template.
      *  If passed as an array, it can be used for template inheritance (first found template "wins").
      *  Usually the array values are PHP class names, which directly correlate to template names.
      *  <code>
      *  array('MySpecificPage', 'MyPage', 'Page')
      *  </code>
-     * @param TemplateParser $parser
-     */
-    public function __construct($templates, TemplateParser $parser = null)
-    {
-        if ($parser) {
-            $this->setParser($parser);
-        }
-
-        $this->setTemplate($templates);
-
-        if (!$this->chosen) {
-            $message = 'None of the following templates could be found: ';
-            $message .= print_r($templates, true);
-
-            $themes = SSViewer::get_themes();
-            if (!$themes) {
-                $message .= ' (no theme in use)';
-            } else {
-                $message .= ' in themes "' . print_r($themes, true) . '"';
-            }
-
-            user_error($message ?? '', E_USER_WARNING);
-        }
-    }
-
-    /**
-     * Triggered early in the request when someone requests a flush.
-     */
-    public static function flush()
-    {
-        SSViewer::flush_template_cache(true);
-        SSViewer::flush_cacheblock_cache(true);
-    }
-
-    /**
-     * Create a template from a string instead of a .ss file
-     *
-     * @param string $content The template content
-     * @param bool|void $cacheTemplate Whether or not to cache the template from string
-     * @return SSViewer
      */
-    public static function fromString($content, $cacheTemplate = null)
+    public function __construct(string|array $templates, ?TemplateEngine $templateEngine = null)
     {
-        $viewer = SSViewer_FromString::create($content);
-        if ($cacheTemplate !== null) {
-            $viewer->setCacheTemplate($cacheTemplate);
+        if ($templateEngine) {
+            $templateEngine->setTemplate($templates);
```

Comment:
I've noticed there's lots of string|array $template ... ->setTemplate() mixing of singulars are plurals happening

Would it be feasible to just force a single type, either string or array?

Reply:
See https://github.com/silverstripe/silverstripe-framework/pull/11405#discussion_r1801846845